Add 'category' and 'aliment' fields to 'Product' table

choogland opened the pull request:
Add 'category' and 'aliment' fields to 'Product' table
- Add 'category' (StrEnum) field to 'Product' table: the normalised category
- Add 'alimment' (str) field to 'Product' table: the aliment (subcategory) of the product 
- Add migration script
- Update mermaid diagram & SVG diagram
- Update unit test accordingly
- Add category/aliment to the parameters passed to the crawler

Commit:
Add 'category' and 'department' fields to 'Product' table

Add 'category' and 'department' fields to 'Product' table
- Add 'category' (StrEnum) field to 'Product' table: the normalised category
- Add 'department' (StrEnum) field to 'Product' table: the store main department
- Add migration script
- Update mermaid diagram & SVG diagram
- Update unit test accordingly

Commit:
Add 'Epicerie sucrée' to Department list

SkypLabs commented on docs/rdb_diagram.mermaid:
```diff
@@ -10,6 +10,10 @@ erDiagram
         enum quantity_unit "Either a weight or a volume"
+
+        enum category "The normalised category of the product"
+        enum department "The store main department of the product"
```
"The main store department".

choogland replied:
Done

SkypLabs commented on src/models/product.py:
```diff
@@ -42,6 +42,74 @@ def _missing_(cls, value):
+@unique
+class Department(StrEnum):
+    """
+    The store main department of the product.
```
"The main store department".

choogland replied:
Done

SkypLabs commented on tests/integration/models/test_product_source.py:
```diff
@@ -69,6 +71,8 @@ def test_create_products(self):
                 quantity=1.5,
                 quantity_unit=QuantityUnit.LITRE,
+                category=Category.LAITIER,
+                department=Department.LAITIER,
```
Could you please test these values?

Something like:

```python
self.assertEqual(result.category, product_a.category)
self.assertEqual(result.department, product_a.department)
```

choogland replied:
Done

SkypLabs commented on tests/unit/models/test_product.py:
```diff
@@ -34,5 +34,97 @@ def test_string_to_enum(self):
+    """
+
+    def test_string_to_enum(self):
+        category = Category("Viandes")
+        self.assertIsInstance(category, Category)
+        self.assertIs(category, Category.VIANDE)
```
Could you please also test the good functioning of the `_missing_` method?

For example, add a few tests as follows:

```python
category = Category("ViAnDes")
self.assertIsInstance(category, Category)
self.assertIs(category, Category.VIANDE)
```

choogland replied:
Done

Commit:
Add category and department to Auchan scraper

- Add category and department scraping and storage to Auchan scraper

Commit:
Fix LINT error

Commit:
Add some other products

SkypLabs commented on src/models/product.py:
```diff
@@ -42,6 +42,74 @@ def _missing_(cls, value):
+    EPICERIE = "Epicerie salée"  # cn06
+    SUCRE = "Epicerie sucrée"  # cn05
+    FRUIT_LEGUME = "Fruits, légumes"  # cn03
+    SURGELE = "Surgelés"  # cn04
```
These are the store departments of Auchan, not our "Aliment" sub-category.

We need to implement our categories and sub-categories as database columns, and then extract and cross match in each scraper the distributor's own categories and ours.

I recommend moving this enumeration in `src/scraper_auchan`.

choogland replied:
Done

SkypLabs commented on migrations/versions/0ea78d81a3aa_add_category_and_department_to_product_.py:
```diff
@@ -0,0 +1,59 @@
+    category_type.create(bind, checkfirst=True)
+
+    department_type = sa.Enum(
+        Department, native_enum=False, values_callable=lambda e: [x.value for x in e]
+    )
+    department_type.create(bind, checkfirst=True)
+
+    op.add_column(
+        "product",
+        sa.Column(
+            "category",
+            category_type,
+        ),
+    )
+    op.add_column(
+        "product",
+        sa.Column(
+            "department",
+            department_type,
+        ),
+    )
```
The `category` and `department` columns aren't optional, so this `upgrade` method needs to update all current products to assign them to a category and a department.

You can take inspiration from:

https://github.com/dataforgoodfr/offseason-proteines-resilientes/blob/2c8b99357a406219da0f33a8f981fb7dea00d429/migrations/versions/d483c555d6d1_add_field_origin_to_source_table.py#L24-L50

Given that we don't know which category and department the previous products need to be placed in, please add an `Unknown` value to both enums and use it in the `upgrade` method here.

choogland replied:
Done

SkypLabs commented on src/models/product.py:
```diff
@@ -67,4 +135,20 @@ class Product(Base):
+        ),
+    )
+
+    department: Mapped["Department"] = mapped_column(
```
`department` is a concept proper to the distributor. It should be renamed to something else.

In the shared spreadsheet, they used the wording "Catégorie" and "aliment", but I don't think `food` would be a good subcategory. When reading the table columns, it wouldn't be obvious that `food` is a subcategory of `category`...

Actually, we could go with `category` and `subcategory`. What do you think?

choogland replied:
I put category and aliment, to align with the excel data to be used.
Department (proper to the store) is moved to the scraper, and used to filter out unwanted products.

Commit:
Updates following PR72 review

- Add some unit tests
- Remove Department from Product table, move Department enum to scraper
- Add aliment (subcategory) to Product table
- Add parameters category and aliment to scraper cli

Commit:
Fix issue on ruff check

Commit:
Update scrapers to store category and aliment

Update all scrapers (Auchan, Biocoop, Carrefour) to store category and aliment

SkypLabs commented:
I just rebased your branch onto `develop`. Starting my review now.

SkypLabs commented:
I just merged my implementation of the categories (b49fd3c). I will now work on adding the `--category` option to the argument parsers and then integrate your work regarding filtering out the undesired products based on their store department.

SkypLabs commented:
I have added the `--category` option to the argument parsers in fdc93cb, 5bc4c3d, and 44c9c2a.

The logic to filter out unwanted products has been added to Auchan and Carrefour respectively in 1442ffd and 19cb059. I doesn't work well for Biocoop as the breadcrumbs are loaded dynamically via JavaScript, so the breadcrumb values are not present in the page processed by Scrapy. I pushed my work-in-progress in this branch if you want to have a look: https://github.com/dataforgoodfr/offseason-proteines-resilientes/tree/feature/filter-out-biocoop.

I'm closing this PR as everything has now been integrated into `develop`. Thanks!
